logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EphemeralMessageDeletionHandler.kt

```diff
@@ -0,0 +1,89 @@
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.datetime.Clock
+import kotlin.coroutines.CoroutineContext
+
+
+interface EphemeralMessageDeletionHandler {
+
+    fun startSelfDeletion(conversationId: ConversationId, messageId: String)
+    fun enqueuePendingSelfDeletionMessages()
+}
+
+
+// TODO:Mateusz: implement failure logic
+internal class EphemeralMessageDeletionHandlerImpl(
+    private val messageRepository: MessageRepository,
+    private val kaliumDispatcher: KaliumDispatcher = KaliumDispatcherImpl,
+    userSessionCoroutineScope: CoroutineScope
+) : EphemeralMessageDeletionHandler, CoroutineScope by userSessionCoroutineScope {
+    override val coroutineContext: CoroutineContext
+        get() = kaliumDispatcher.default
+
+    private val onGoingSelfDeletionMessagesMutex = Mutex()
+
+    private val onGoingSelfDeletionMessages = mutableMapOf<Pair<ConversationId, String>, Unit>()
+    override fun startSelfDeletion(conversationId: ConversationId, messageId: String) {
+        launch {
+            onGoingSelfDeletionMessagesMutex.withLock {
+                val isSelfDeletionOutgoing = onGoingSelfDeletionMessages[conversationId to messageId] != null
+                if (isSelfDeletionOutgoing) return@launch
+
+                onGoingSelfDeletionMessages[conversationId to messageId] = Unit
+            }
+
+            messageRepository.getMessageById(conversationId, messageId).map { message ->
+                if (message is Message.Regular) enqueueSelfDeletion(message = message)
+            }
+        }
+    }
+
+    private suspend fun enqueueSelfDeletion(message: Message.Regular) {
+        message.expirationData?.let { expirationData ->
+            with(expirationData) {
+                if (selfDeletionStatus is Message.ExpirationData.SelfDeletionStatus.NotStarted) {
+                    messageRepository.markSelfDeletionStartDate(
+                        conversationId = message.conversationId,
+                        messageUuid = message.id,
+                        deletionStartDate = Clock.System.now()
+                    )
+                }
+
+                delay(timeLeftForDeletion())
+            }
+
+            onGoingSelfDeletionMessagesMutex.withLock {
+                onGoingSelfDeletionMessages - message.conversationId to message.id
+            }
+
+            messageRepository.deleteMessage(message.id, message.conversationId)
+        }
+    }
+
+    override fun enqueuePendingSelfDeletionMessages() {
+        launch {
+            messageRepository.getEphemeralMessagesMarkedForDeletion()
+                .onSuccess { ephemeralMessages ->
+                    ephemeralMessages.forEach { ephemeralMessage ->
+                        launch {
+                            if (ephemeralMessage is Message.Regular) {
+                                enqueueSelfDeletion(message = ephemeralMessage)
+                            }
+                        }
```

Comment:
I would expect that the function `enqueueSelfDeletion` would have a `launch` or similar inside of it anyway. _i.e._ no need to wrap it around a launch.

Can we move this `launch` block to inside the `enqueueSelfDeletion` function and keep the same behaviour similar between these?

Reply:
I do not know why I missed that :S yes thank u